### DISCUSSION TITLE
Kitty + Tmux + LF + Image preview!

### DISCUSSION THREAD

**Discussion by jackielii** [gokcehan/lf#1153] [Show and tell]
Today I tried out lf with:

kitty: nightly at https://github.com/kovidgoyal/kitty/commit/31d8a98a45a214a361e38f9e52a13a17faabfdb4
tmux: 3.3a

It looks like this

https://user-images.githubusercontent.com/360983/224049763-b2d5dcec-3cf0-485b-90d7-39bf4f59deb6.mov

Mostly working great!

A couple of issues still exists:

1. When switching to the image, the font changes (from 22s in the clip). I suspect this is an issue in kitty nightly.
    Somehow kitty nightly increases the font weight by 1. I.e. I use to use `Fira Code Medium`, but now I have to use `Fira Code Regular` to have the same effect
1. "loading..." text is not cleared and overlays on top of the image
2. The preview doesn't clear properly. The info is ok through. (See my dotfiles[1] for the mapping)
3. I had to decrease the size by one in order to have the image fit the screen. Maybe it's because lf didn't take the border into account? Also see [my dot files](https://github.com/jackielii/dotfiles/blob/main/.config/lf/preview) for details

**Comment by madrigal1** [comment]
Hey great setup !. Also slightly off topic, how which software did you use to record your screen and get it in video form like that ?

**Reply by jackielii** [reply]
oh. it's just macos built in screenshot app

**Comment by Archie-2021** [comment]
Is there a script or something for this. I have heard it is possible to have image preview in `lf` using kitty. But can't find anything.

**Reply by jackielii** [reply]
checkout `preview` or `preview.go` in https://github.com/jackielii/dotfiles/tree/main/.config/lf

I use `preview.go` mostly. The key is to make sure you use the latest version of kitty & tmux

**Comment by Archie-2021** [comment]
> checkout `preview` or `preview.go` in https://github.com/jackielii/dotfiles/tree/main/.config/lf
> 
> I use `preview.go` mostly. The key is to make sure you use the latest version of kitty & tmux

Thanks, does that preview script use pistol? After I asked here I found out about pistol. But I would like to not use it as the package on aur is outofdate. Also is gorun need installing to run go version of the script?

**Reply by jackielii** [reply]
I started with the `preview` script, it doesn't use pistol. But I ended up with my `preview.go` which uses pistol for preview some files, maybe mostly archive files at this point. I plan to replace most of them with my own script anyways.

The `preview` script has most of the stuff you need. IIRC, I adapted from https://github.com/Naheel-Azawy/stpv/blob/master/stpv. But as my changes grew, I realise I'm not very good at bash/zsh scripting, so I moved to a `gorun` script.

I think if you want it to work for you, I suggest you take the same path: you'll eventually want to modify it to your env. So adapt it to something you're comfortable is important.

My preview.go script added a few improvements like generating previews for most files on mac: better than built-in, drawio, video, doc, docx, ppt, word docs etc, also fixed the flakiness when preview process is happening but you're inputing characters etc. Otherwise it was almost line to line adaptation.

**Reply by jackielii** [reply]
btw, my script is for macos, e.g. it uses `qlmanage` to generate some previews

**Comment by Archie-2021** [comment]
> I started with the `preview` script, it doesn't use pistol. But I ended up with my `preview.go` which uses pistol for preview some files, maybe mostly archive files at this point. I plan to replace most of them with my own script anyways.
> 
> The `preview` script has most of the stuff you need. IIRC, I adapted from https://github.com/Naheel-Azawy/stpv/blob/master/stpv. But as my changes grew, I realise I'm not very good at bash/zsh scripting, so I moved to a `gorun` script.
> 
> I think if you want it to work for you, I suggest you take the same path: you'll eventually want to modify it to your env. So adapt it to something you're comfortable is important.
> 
> My preview.go script added a few improvements like generating previews for most files on mac: better than built-in, drawio, video, doc, docx, ppt, word docs etc, also fixed the flakiness when preview process is happening but you're inputing characters etc. Otherwise it was almost line to line adaptation.



> I started with the `preview` script, it doesn't use pistol. But I ended up with my `preview.go` which uses pistol for preview some files, maybe mostly archive files at this point. I plan to replace most of them with my own script anyways.
> 
> The `preview` script has most of the stuff you need. IIRC, I adapted from https://github.com/Naheel-Azawy/stpv/blob/master/stpv. But as my changes grew, I realise I'm not very good at bash/zsh scripting, so I moved to a `gorun` script.
> 
> I think if you want it to work for you, I suggest you take the same path: you'll eventually want to modify it to your env. So adapt it to something you're comfortable is important.
> 
> My preview.go script added a few improvements like generating previews for most files on mac: better than built-in, drawio, video, doc, docx, ppt, word docs etc, also fixed the flakiness when preview process is happening but you're inputing characters etc. Otherwise it was almost line to line adaptation.

Ok I get it now. I had a problem before getting these preview script to work, now it makes more sense. I tried your script and it freezes on first pdf file and stay that way. I guess it is because of your mac setup as you said. but I could not notice the path to your `pistol.conf` (in your shell script) is to kitty directory not the one in lf directory. can you explain about that?

I think I would start to play around with that script and add stuff step by step. I like bash scripting altho I am new to it, so it would be a fun project right now image previews and archive files are fine . but I need a lot more to add.

Thanks

**Reply by emersonfbarros** [reply]
nowadays in the latest versions of tmux, kitty and lf the [ctpv](https://github.com/NikitaIvanovV/ctpv) preview works out of the box, but the cleaner doesn't work, I tried to make it work without success so far 😢

**Comment by jphalip** [comment]
Hi @jackielii - Sorry to ping you so long after the last reply in this thread, but just in case you're still working on this, I was curious if you had any updates.

I've tried to follow the examples from your [dotfiles](https://github.com/jackielii/dotfiles/blob/main/.config/lf/lfrc) by copying your go files, installing `gorun`, and adding the following to my `lfrc`:

```
set previewer ~/.config/lf/preview.go
set cleaner ~/.config/lf/clean.go
```

What happens is that when I browse files from a directory, it seems to cache the first preview and then displays the same image for all other files. And if I try it in Tmux then the same thing happens, except it briefly appears then disappears.

I'm using Tmux v3.5a and lf v33.

If you have any tips, please let me know. Thanks in advance!
